File: examples/MPI_Benchmark/source_code/pingping/src/python/pingping.py
#!/usr/bin/env python
# -*- coding: latin-1 -*-

import sys
import datetime
from threading import Thread

class ProcPing(Thread):
	def __init__(self, name, data, qtdMsg):
		Thread.__init__(self)							
		self.name = name
		self.data = data
		self.qtdMsg = qtdMsg
		self.mailBox = []

	def setPeer(self, Peer):
		self.Peer = Peer

	def send(self, dado):		
		self.Peer.mailBox = dado

	def recv(self):
		while True:	
			if not len(self.mailBox) < len(self.data):								
				print(self)				
				self.mailBox = []
				break

	def run(self):
		for i in range (0, self.qtdMsg + 1):
			self.send(self.data)
			if i < self.qtdMsg:
				self.recv()

class PingPing(Thread):

	def __init__(self, tamMsg, qtdMsg):
		Thread.__init__(self)
		self.tamMsg = tamMsg
		self.qtdMsg = qtdMsg

	def run(self):
		index = 0
		array = [1]			
		while index < self.tamMsg -1:
			array.append(1)
			index = index + 1
		
		p1 = ProcPing("1", array, self.qtdMsg)
		p2 = ProcPing("2", array, self.qtdMsg)
		
		p2.setPeer(p1)
		p1.setPeer(p2)
		
		timeStart = datetime.datetime.now()
		p1.start()		
		p2.start()
		
		p1.join()
		p2.join()
		timeEnd = datetime.datetime.now()
		
		timeExec = timeEnd - timeStart

		line = "%d\t%d\t%s\n" % (self.tamMsg, self.qtdMsg, timeExec)
	
		try:
			arq = open('saida.txt', 'r')
			textoSaida = arq.read()
			arq.close()
		except:
			arq = open('saida.txt', 'w')
			textoSaida = ""
			arq.close()

		arq = open('saida.txt', 'w')
		textoSaida = textoSaida + line
		arq.write(textoSaida)
		arq.close()

def main():
	param = sys.argv[1:]

	tamMsg = int(param[0])
	qtdMsg = int(param[1])

	pingPing = PingPing(tamMsg, qtdMsg)
	pingPing.start()

if __name__=="__main__":
	main()

File: examples/MPI_Benchmark/source_code/pingping/java/PingPing.java
public class PingPing extends Thread{

	private final int QTD_PROC_TOTAL = 2;
	private int qtdProcTerminados = 0;
	
	private long timeSpawn;
	private long timeExec;

	public long timeStart;
	public long timeEnd;

	private int tamDados;
	private int qtdRept;
	private String outLocation;
	
	public PingPing(int tamDados, int qtdMsg, String outLocation) {
		this.tamDados = tamDados;
		this.qtdRept = qtdMsg;
		this.outLocation = outLocation;
	}

	public void run() {
		byte[] dado = generateData(tamDados);

		timeStart = timeMicroSeg();
		ProcPing p1 = new ProcPing("1", dado, this, qtdRept);
		ProcPing p2 = new ProcPing("2", dado, this, qtdRept);
		timeEnd = timeMicroSeg();

		timeSpawn = timeEnd - timeStart;
		
		timeStart = timeMicroSeg();
		p1.setPeer(p2);
		p1.start();
		p2.setPeer(p1);
		p2.start();
		
		dormirAteTerminar();

		timeEnd = timeMicroSeg();

		timeExec = timeEnd - timeStart;
		
		Salvar.writeResultPeer(outLocation, tamDados, qtdRept, timeExec, timeSpawn);
	}

	private synchronized void dormirAteTerminar() {
		while(qtdProcTerminados!= QTD_PROC_TOTAL){
			try {
				wait();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}		
	}

	public synchronized void acordar() {
		qtdProcTerminados++;
		notifyAll();
	}

	private byte[] generateData(int tamDados) {
		byte[] dado = new byte[tamDados];
		return dado;
	}

	private long timeMicroSeg() {
		return System.nanoTime()/1000;
	}
}
